File: algorithm_library/python/math/modComb.py
# 二項係数ライブラリ

# 前計算 O(n)
# O(1) で nCk, nHk を返す
class Comb:
  def __init__(self,n,mod):
    self.mod = mod
    self.n = n
    self.fact = [1, 1] 
    self.factinv = [1, 1]  
    self.inv = [0, 1] 
    for i in range(2,self.n+1):
        self.fact.append((self.fact[i-1] * i) % self.mod)
        self.inv.append((self.mod-self.inv[self.mod % i] * (self.mod // i)) % self.mod)
        self.factinv.append((self.factinv[i-1] * self.inv[i]) % self.mod)
        
  def C(self,n,k): # nCk % modを返す
    assert n <= self.n
    if (k < 0 or n < 0) or (n < k):return 0
    if k==1: return n
    k = min(k,n-k)
    return self.fact[n]*self.factinv[k]%self.mod*self.factinv[n-k]%self.mod

  # 重複組合せ nHr % modを返す
  def H(self,n,r):
      return self.C(n+r-1,r)

# O(k) 前計算
# O(k) で nCk を計算（n が大きいときでも使える）
class Comb:
    def __init__(self, k, mod):
        self.k = k
        self.mod = mod
        self.fact_inv = [0] * (k + 1)
        self.inv = [0] * (k + 1)
        self.fact_inv[0] = 1
        if k >= 1:
            self.fact_inv[1] = 1
            self.inv[1] = 1

        for i in range(2, k + 1):
            self.inv[i] = mod - self.inv[mod % i] * (mod // i) % mod
            self.fact_inv[i] = self.fact_inv[i - 1] * self.inv[i] % mod

    def nCk(self, n, k):
        if k < 0 or k > n:
            return 0
        if k > self.k:
            raise ValueError("precomputed k is too small")
        ans = 1
        for i in range(k):
            ans *= n - i
            ans %= self.mod
        return ans * self.fact_inv[k] % self.mod

File: algorithm_library/cpp-from-py/math/modComb.cpp
#include <bits/stdc++.h>
using namespace std;

struct Comb {
    int k;
    long long mod;
    vector<long long> fact_inv, inv;

    Comb(int k_, long long mod_) : k(k_), mod(mod_), fact_inv(k + 1, 0), inv(k + 1, 0) {
        fact_inv[0] = 1;
        if (k >= 1) fact_inv[1] = 1, inv[1] = 1;
        for (int i = 2; i <= k; i++) {
            inv[i] = mod - inv[mod % i] * (mod / i) % mod;
            fact_inv[i] = fact_inv[i - 1] * inv[i] % mod;
        }
    }

    long long nCk(long long n, int kk) const {
        if (kk < 0 || kk > n) return 0;
        if (kk > k) throw invalid_argument("precomputed k is too small");
        long long ans = 1;
        for (int i = 0; i < kk; i++) ans = ans * ((n - i) % mod) % mod;
        return ans * fact_inv[kk] % mod;
    }
};

struct CombAll {
    int n;
    long long mod;
    vector<long long> fact, factinv, inv;

    CombAll(int n_, long long mod_) : n(n_), mod(mod_), fact{1, 1}, factinv{1, 1}, inv{0, 1} {
        for (int i = 2; i <= n; i++) {
            fact.push_back(fact[i - 1] * i % mod);
            inv.push_back((mod - inv[mod % i] * (mod / i) % mod) % mod);
            factinv.push_back(factinv[i - 1] * inv[i] % mod);
        }
    }

    long long C(int nn, int kk) const {
        assert(nn <= n);
        if (kk < 0 || nn < 0 || nn < kk) return 0;
        if (kk == 1) return nn;
        kk = min(kk, nn - kk);
        return fact[nn] * factinv[kk] % mod * factinv[nn - kk] % mod;
    }

    long long H(int nn, int r) const {
        return C(nn + r - 1, r);
    }
};
